doc: add AI guidelines (#17365)
rook/rook: Documentation/Contributing/ai-guidelines.md

---

> title: AI Guidelines

Reviewer: According to the `.pages` file, this new topic will show up in the TOC after the documentation (third) topic. How about we put it earlier in the list?

Author: I'll plan to make it the 2nd topic. Development flow seems like the right first topic to me

---

> First and foremost, Rook's stance on AI guidelines is focused on preserving trust. Thousands of Rook users trust Rook and its maintainers to release enterprise-grade software with as few issues as possible and with as much transparency as possible. Each person who contributes code to Rook must be able to do so in a way that preserves maintainer trust, and by extension, user trust.

Reviewer: ```suggestion
Rook's AI guidelines are focused on preserving trust. Thousands of Rook users trust Rook and its maintainers to release enterprise-grade software with as few issues as possible and with as much transparency as possible. Each person who contributes code to Rook must be able to do so in a way that preserves maintainer trust, and by extension, user trust.
```

## Legal requirements

As a CNCF project, Rook follows the [Linux Foundation Guidelines](https://www.linuxfoundation.org/legal/generative-ai) for ensuring legal compliance. These guidelines require contributors to check (in summary):

1. AI tool restrictions: Contributors must ensure the AI tool's terms don't impose contractual limitations that conflict with the project's license.
2. Third-party content permissions: If generated output contains copyrighted materials from others, contributors must confirm proper permissions exist (via compatible license) and provide license information when contributing.

In order to ensure legal guidelines, Rook requires human involvement and engagement and does not allow fully-autonomous or near-fully-autonomous "bots" to submit issues or pull requests. If Rook maintainers have reasonable cause to believe a contribution does not have human oversight, it will be rejected.

## Transparency requirements

> 1. All contributors should disclose AI tool use in pull request descriptions when submitting changes.

Reviewer 2: Should we provide an example of acceptable ways to make this disclosure?

I don't have strong feelings about how it is done other than I would dislike the disclosure as code comments.

Reviewer: I struggle with this ask. What do we gain by knowing the tool they used? To me, that doesn't help with the fundamental goal to ensure all code is understood, clean, and validated with thorough testing. 
Maybe instead of asking which AI tool, we ask them to describe how AI contributed to the PR, for example which large chunks of code might have been generated.

Author: This is a good point. Also FWIW, this is just copied from I think KubeVirt's policy. I don't particularly care what tool Rook contributors use, but as a reviewer I would like to understand when AI has generated code that's more than maybe 10-ish lines.

Update: I updated the wording to try to clarify that the proposal is more about what changed rather than which tool.

`Contributors should disclose in pull request descriptions how AI tools assisted in development. E.g., generating substantial functions, documentation, and/or unit tests.`

I think more refinement could be good as well.

> 2. Rook does not require or prohibit the use of commit message "trailer" lines which identify the LLM(s) used.

Reviewer: I kind of prefer the [Kubernetes policy](https://www.kubernetes.dev/docs/guide/pull-requests/#ai-guidance) that says those properties are not to be specified for AI tools. Maybe we consider that? It's the human that is on the hook and should sign it.

Author: I think I tend to agree here also. I think the human signoff is important, and it's telling that the k/k project went this route. In my mind human-only signoff reinforces the requirement to have human-driven contributions, and I suspect that's why they went this route.

When I drafted the policy, it seemed like having no preference would be easiest for us to police, which might reduce our maintainer burden a tiny amount.

I'm really curious to hear other thoughts here as well. I would prefer not require these signoffs, but I am open to either of the two alternatives: require no LLM signoff, or make no requirement.

Update: I adjusted the second draft to use Kubernetes's policy of AI tool signoffs not being allowed, while still hoping to continue with feedback discussions

> ## Please respect maintainer time

Reviewer: How about this title? We aren't going to necessarily reject everything that doesn't follow these guidelines, so they seem like best practices.
```suggestion
## Best practices to respect maintainer time
```

> 2. Rook requires human sign-off. Signing commits using an AI tool, or listing AI tooling in commits using `co-authored-by`, `assisted-by`, `generated-by`, or using similar commit trailers is not allowed.

Reviewer: What about this? As long as a human primarily signs off, perhaps the other tags don't matter so much.
```suggestion
2. Rook requires human sign-off. Signing commits using an AI tool, or listing AI tooling in commits using `co-authored-by`, `assisted-by`, `generated-by`, or using similar commit trailers is not necessary.
```

> AI models are making it much easier to contribute code to open source, and maintainer time is limited. Please respect the time of Rook maintainers by:

Reviewer: IMO it sounds better if we just state the guidelines and respect will come implicitly.
```suggestion
AI models are making it much easier to contribute code to open source, and maintainer time is limited. 
```

> 1. Opening issues with clear proposals before starting work not already outlined in an existing issue.

Reviewer: How is this AI related? Perhaps we include this as a recommendation in the main development guide instead of this doc.

Author: It isn't truly AI-related in most ways. I think you are right that it belongs in the main dev guide, and I'll plan to make that change.

I do have one doubt still. I think we are seeing the pattern that new contributors empowered by AI are creating PRs without issues, or creating issues then immediately creating PRs, with more frequency than before.

I think it could be a good change to add this to the dev guide, and also retain a short reminder here with a link to the longer policy. WDYT?

Reviewer: I would actually prefer not to document a process around opening issues before PRs. It is certainly a good practice to open an issue if there are questions to address before opening a PR, but if there is already confidence in a bug fix or small feature, opening an issue seems like overhead to me. PRs and issues are both open for comments, so they are equivalent in many ways. 

Perhaps the process we care more about is that for larger features, design doc PRs should be opened and merged before PRs to implement are reviewed. Even for large features, I don't see an issue as strictly necessary.

Author: I am still on the fence, but I think I see where you are coming from.

I agree with you especially when PRs come from frequent contributors. I also don't like unnecessary overhead, and I don't have concerns with small to medium sized PRs coming from maintainers without an issue first.

What I have seen, which is behind my desire, is many new contributors showing up unexpectedly with non-trivial PRs that can be a burden to understand and review -- even when they are very confident that what they did is good. I suspect it is a poor experience for them when we ask them to open an issue first, or adjust their entire approach after writing code. In theory, if new contributors follow an issue-first policy, we can help make sure things go smoothly for them. 

However, maybe this guideline is a net zero. I might find that in reality, this policy is no less burden than the status quo today.

I think we could refine the policy further, if we think this has some benefit. We could say that any PR over a certain size (let's say 30ish added lines of code) needs an issue and discussion first. That could still allow for simple doc updates and bug fixes without unnecessary overhead.

We can also drop this. It's definitely not a priority change.

Reviewer: As a general guideline, I'm ok to specify it as written. I just don't anticipate many new users to read the policies before they open PRs, so it may not change behavior much.

> 2. Starting with small pull requests related to single issues (one at a time for new contributors).

Reviewer: Why this recommendation? Is the point that we will likely reject large feature PRs from first-time users?

Author: These are fair questions. This was a pretty lazy copy from OPA's policy on my part.

For me, I think the impactful thing to maintainers is not opening large feature PRs driven by AI that are a nightmare to review.

It might be less confusing to link to k/k's policy of not allowing [large or automatic AI edits](https://www.kubernetes.dev/docs/guide/pull-requests/#large-or-automatic-edits).

I think the muddy part is that I would be fine with large AI PRs that do simple things like updating dependencies or refactoring a specific golang usage pattern. I considered adopting KubeVirt's [acceptable uses of AI tools](https://github.com/kubevirt/community/blob/main/ai-contribution-policy.md#acceptable-uses-of-ai-tools) section, but this seemed like it might be a little over-specified.

In the interest of not over-specifying, I wonder how everyone feels about linking to k/k's no-large-AI-edits policy in lieu of this.

Reviewer: Sounds great to link to that K8s policy on large or automatic edits. It sounds like exactly what we are concerned about as well.

> 2. Do not submit large AI-generated PRs without approval. Follow [Kubernetes's guidelines for large/automatic edits](https://www.kubernetes.dev/docs/guide/pull-requests/#large-or-automatic-edits) as needed.

Reviewer: ```suggestion
2. Do not submit large AI-generated PRs without previous design discussion. Follow [Kubernetes's guidelines for large/automatic edits](https://www.kubernetes.dev/docs/guide/pull-requests/#large-or-automatic-edits) as needed.
```

> 3. Not using LLM output to write commit messages or GitHub comments. Reviewers are interested in knowing your reasoning about the code you submitted for review. Even if an LLM helped you write that code, it's yours to own and explain. Engaging an AI in discussions with human community members who took the time to review your code and provide you with personal feedback is considered disrespectful and may result in a PR being rejected.

Reviewer: I like how the Kubernetes policy just says simply that they want to interact with the person. Maybe we can just simplify to something like this?
```suggestion
3. Do not use LLM output to write commit messages or GitHub comments. Reviewers are interested in knowing your reasoning about the code you submitted for review. It's yours to own and explain. Automated responses may result in a PR being rejected.
```

Author: Thanks for this. I agree. I was struggling to reduce it more
